keps/369-job-interface/README.md

```diff
@@ -0,0 +1,423 @@
+  PodSets() []kueue.PodSet
+  // EquivalentToWorkload validates whether the workload is semantically equal to the job.
+  EquivalentToWorkload(wl kueue.Workload) bool
+  // PriorityClass returns the job's priority class name.
+  PriorityClass() string
+  // QueueName returns the queue name the job enqueued.
+  QueueName() string
+  // Ignored instructs whether this job should be ignored in reconciling, e.g. lacking the queueName.
+  Ignored() bool
+}
+```
+
+Also defined another interface `SequenceJob` for scheduling in sequence requirements, see design details here [All-or-nothing semantics for job resource assignment](https://github.com/kubernetes-sigs/kueue/tree/main/keps/349-all-or-nothing#kep-349-all-or-nothing-semantics-for-job-resource-assignment).
+
+```golang
+type SequenceJob interface {
+    // PodsReady instructs whether job derived pods are all ready now.
+    PodsReady() bool
+}
+```
+
+Part II: Kubernetes Job implementation details
+
+We'll wrap the batchv1.Job to `BatchJob` who implements the GenericJob interface, as well as SequenceJob interface.
+
+```golang
+type BatchJob struct {
+  batchv1.Job
+}
+
+var _ GenericJob = &BatchJob{}
+var _ SequenceJob = &BatchJob{}
+```
+
+Part III: Library
+
+We'll have several helper functions in constructing controllers, they can be imported directly in the future:
+
+```golang
+// EnsureOneWorkload will query for the single matched workload corresponding to job and return it.
+// If there're more than one workload, we should delete the excess ones.
+// The returned workload could be nil.
+EnsureOneWorkload(ctx context.Context, cli client.Client, req ctrl.Request, record record.EventRecorder, job GenericJob) (*kueue.Workload, error)
+
+// StartJob will unsuspend the job, and also inject the node affinity.
+StartJob(ctx context.Context, client client.Client, wl *kueue.Workload, condition metav1.Condition) error
+
+// StopJob will suspend the job, and also restore node affinity, reset job status if needed.
+ StopJob(ctx context.Context, client client.Client, job GenericJob, wl *kueue.Workload) error
+
+ // CreateWorkload will derive and create a workload from the corresponding job.
+ CreateWorkload(ctx context.Context, client client.Client, scheme *runtime.Scheme, job GenericJob) (*kueue.Workload, error)
+
+ // GetNodeSelectors will extract node selectors from admitted workloads.
+ GetNodeSelectors(ctx context.Context, client client.Client, w *kueue.Workload) ([]map[string]string, error)
+
+ // UpdateQueueNameIfChanged will update workload queue name if changed.
+ UpdateQueueNameIfChanged(ctx context.Context, client client.Client, job GenericJob, wl *kueue.Workload) error
+
+ // SetWorkloadCondition will update the workload condition by the provide one.
+ SetWorkloadCondition(ctx context.Context, client client.Client, wl *kueue.Workload, condition metav1.Condition) error
+```
+
+Part IV: Working Flow in pseudo-code
+
+```golang
+Reconcile:
```

Comment:
I think this should be part of the library. Something along these lines:

```golang
type Controller[J GenericJob] struct {}

func NewController[J GenericJob]() Controller[J] {
  return Controller[J]{}
}
```

Reply:
Some hooks maybe needed, but we can add them later.

Reply:
The disadvantage is the interface may be changed more frequently. We can make the interface as small as possible to mitigate this.